phal: Hardware procedure error handling support

aravynd opened the pull request:
Test result:

root@xxx:~# peltool -i 0x50000842
{
"Private Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc host processor control",
    "Created at":               "12/02/2025 06:38:08",
    "Committed at":             "12/02/2025 06:38:08",
    "Creator Subsystem":        "BMC",
    "CSSVER":                   "",
    "Platform Log Id":          "0x50000842",
    "Entry Id":                 "0x50000842",
    "BMC Event Log Id":         "753"
},
"User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Log Committed by":         "bmc error logging",
    "Subsystem":                "CEC Hardware",
    "Event Scope":              "Entire Platform",
    "Event Severity":           "Unrecoverable Error",
    "Event Type":               "Not Applicable",
    "Action Flags": [
                                "Service Action Required",
                                "Report Externally",
                                "HMC Call Home"
    ],
    "Host Transmission":        "Not Sent",
    "HMC Transmission":         "Not Sent"
},s
"Primary SRC": {
    "Section Version":          "1",
    "Sub-section type":         "1",
    "Created by":               "bmc host processor control",
    "SRC Version":              "0x02",
    "SRC Format":               "0x55",
    "Virtual Progress SRC":     "False",
    "I5/OS Service Event Bit":  "False",
    "Hypervisor Dump Initiated":"False",
    "Backplane CCIN":           "2E2D",
    "Terminate FW Error":       "False",
    "Deconfigured":             "False",
    "Guarded":                  "False",
    "Error Details": {
        "Message":              "Failure occurred during boot process"
    },
    "Valid Word Count":         "0x09",
    "Reference Code":           "BD503001",
    "Hex Word 2":               "00000055",
    "Hex Word 3":               "2E2D0010",
    "Hex Word 4":               "00000000",
    "Hex Word 5":               "00000000",
    "Hex Word 6":               "00000000",
    "Hex Word 7":               "00000000",
    "Hex Word 8":               "00000000",
    "Hex Word 9":               "00000000"
},
"Extended User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc error logging",
    "Reporting Machine Type":   "9105-22B",
    "Reporting Serial Number":  "139F210",
    "FW Released Ver":          "",
    "FW SubSys Version":        "fw1210.00-1.21",
    "Common Ref Time":          "00/00/0000 00:00:00",
    "Symptom Id Len":           "20",
    "Symptom Id":               "BD503001_2E2D0010"
},
"Failing MTMS": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc error logging",
    "Machine Type Model":       "9105-22B",
    "Serial Number":            "139F210"
},
"User Data 0": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "BMCLoad": "1.54 1.16 1.15",
    "BMCState": "Ready",
    "BMCUptime": "0y 0d 1h 44m 17s",
    "BootState": "Unspecified",
    "ChassisState": "TransitioningToOn",
    "FW Version ID": "fw1210.00-1.21-3-g1df0c30c21-dirty",
    "HostState": "TransitioningToRunning",
    "System IM": "50001001"
},
"User Data 1": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "HWP_CDG_TGT_01_CO_PRIORITY": "MEDIUM",
    "HWP_CDG_TGT_01_CO_REQ": "true",
    "HWP_CDG_TGT_01_DECONF_REQ": "true",
    "HWP_CDG_TGT_01_GUARD_REQ": "true",
    "HWP_CDG_TGT_01_GUARD_TYPE": "GARD_Fatal",
    "HWP_FFDC_AVS_BUS": "01",
    "HWP_FFDC_AVS_RAIL": "00",
    "HWP_FFDC_CBS_ENVSTAT_READ": "3c000020",
    "HWP_HW_CO_01_CALLOUT_PLANAR": "false",
    "HWP_HW_CO_01_CLK_POS": "255",
    "HWP_HW_CO_01_HW_ID": "UN_SUPPORTED_PART",
    "HWP_HW_CO_01_PRIORITY": "HIGH",
    "HWP_RC": "RC_VDN_PGOOD_NOT_SET",
    "HWP_RC_DESC": "Nest power (VDN) Power Good indication not set"
},
"User Data 2": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "Data": [
        {
            "Deconfigured": true,
            "EntityPath": [
                35,
                1,
                0,
                2,
                0,
                5,
                0
            ],
            "GuardType": "GARD_Fatal",
            "Guarded": true,
            "Priority": "M"
        }
    ]
},
"User Data 3": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "PEL Internal Debug Data": {
        "SRC": [
            "Failed extracting callout data from JSON: JSON callout needs either an inventory path or location code"
        ]
    }
}
}

Change-Id: I4feafbbd38eaa1971abf32a37901c498e2e34b73

gcpin reviewed (commented):
please add Signed-off-by field to commit msg.

gcpin commented on error/sbe_hwp_failure.C:
```diff
@@ -0,0 +1,26 @@
+    std::unordered_map<std::string, std::string> effdcUserData;
+    errl::util::convertFAPItoPELformat(ffdc, callout, effdcUserData);
+
+    std::string errMsg = "org.open_power.PHAL.Error.Boot";
```
can be made static constexpr const char* in the header file and be used?

gcpin commented on error/errl_factory.C:
```diff
@@ -10,6 +11,12 @@ std::optional<std::unique_ptr<ErrlHandle>>
     return errl::sbe_boot_failure::create(target);
 }
+
+std::optional<std::unique_ptr<ErrlHandle>> createSbeHWPFailure()
```
Instead of having different functions for getting error handle, can we have a single getErrorHandle method with enum type as argument for different failures?

gcpin reviewed (commented):
please use ``` (triple backticks) to enclose test results in the PR summary.

aravynd committed:
phal: Hardware procedure error handling support

This change is part of the HWP error handling
support and handles the changes in the IPL repo.
The change includes creation of error handle for
a HWP failure during boot.

Test result:
```
root@xxx:~# peltool -i 0x50000842
{
"Private Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc host processor control",
    "Created at":               "12/02/2025 06:38:08",
    "Committed at":             "12/02/2025 06:38:08",
    "Creator Subsystem":        "BMC",
    "CSSVER":                   "",
    "Platform Log Id":          "0x50000842",
    "Entry Id":                 "0x50000842",
    "BMC Event Log Id":         "753"
},
"User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Log Committed by":         "bmc error logging",
    "Subsystem":                "CEC Hardware",
    "Event Scope":              "Entire Platform",
    "Event Severity":           "Unrecoverable Error",
    "Event Type":               "Not Applicable",
    "Action Flags": [
                                "Service Action Required",
                                "Report Externally",
                                "HMC Call Home"
    ],
    "Host Transmission":        "Not Sent",
    "HMC Transmission":         "Not Sent"
},s
"Primary SRC": {
    "Section Version":          "1",
    "Sub-section type":         "1",
    "Created by":               "bmc host processor control",
    "SRC Version":              "0x02",
    "SRC Format":               "0x55",
    "Virtual Progress SRC":     "False",
    "I5/OS Service Event Bit":  "False",
    "Hypervisor Dump Initiated":"False",
    "Backplane CCIN":           "2E2D",
    "Terminate FW Error":       "False",
    "Deconfigured":             "False",
    "Guarded":                  "False",
    "Error Details": {
        "Message":              "Failure occurred during boot process"
    },
    "Valid Word Count":         "0x09",
    "Reference Code":           "BD503001",
    "Hex Word 2":               "00000055",
    "Hex Word 3":               "2E2D0010",
    "Hex Word 4":               "00000000",
    "Hex Word 5":               "00000000",
    "Hex Word 6":               "00000000",
    "Hex Word 7":               "00000000",
    "Hex Word 8":               "00000000",
    "Hex Word 9":               "00000000"
},
"Extended User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc error logging",
    "Reporting Machine Type":   "9105-22B",
    "Reporting Serial Number":  "139F210",
    "FW Released Ver":          "",
    "FW SubSys Version":        "fw1210.00-1.21",
    "Common Ref Time":          "00/00/0000 00:00:00",
    "Symptom Id Len":           "20",
    "Symptom Id":               "BD503001_2E2D0010"
},
"Failing MTMS": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "bmc error logging",
    "Machine Type Model":       "9105-22B",
    "Serial Number":            "139F210"
},
"User Data 0": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "BMCLoad": "1.54 1.16 1.15",
    "BMCState": "Ready",
    "BMCUptime": "0y 0d 1h 44m 17s",
    "BootState": "Unspecified",
    "ChassisState": "TransitioningToOn",
    "FW Version ID": "fw1210.00-1.21-3-g1df0c30c21-dirty",
    "HostState": "TransitioningToRunning",
    "System IM": "50001001"
},
"User Data 1": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "HWP_CDG_TGT_01_CO_PRIORITY": "MEDIUM",
    "HWP_CDG_TGT_01_CO_REQ": "true",
    "HWP_CDG_TGT_01_DECONF_REQ": "true",
    "HWP_CDG_TGT_01_GUARD_REQ": "true",
    "HWP_CDG_TGT_01_GUARD_TYPE": "GARD_Fatal",
    "HWP_FFDC_AVS_BUS": "01",
    "HWP_FFDC_AVS_RAIL": "00",
    "HWP_FFDC_CBS_ENVSTAT_READ": "3c000020",
    "HWP_HW_CO_01_CALLOUT_PLANAR": "false",
    "HWP_HW_CO_01_CLK_POS": "255",
    "HWP_HW_CO_01_HW_ID": "UN_SUPPORTED_PART",
    "HWP_HW_CO_01_PRIORITY": "HIGH",
    "HWP_RC": "RC_VDN_PGOOD_NOT_SET",
    "HWP_RC_DESC": "Nest power (VDN) Power Good indication not set"
},
"User Data 2": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "Data": [
        {
            "Deconfigured": true,
            "EntityPath": [
                35,
                1,
                0,
                2,
                0,
                5,
                0
            ],
            "GuardType": "GARD_Fatal",
            "Guarded": true,
            "Priority": "M"
        }
    ]
},
"User Data 3": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "bmc error logging",
    "PEL Internal Debug Data": {
        "SRC": [
            "Failed extracting callout data from JSON: JSON callout needs either an inventory path or location code"
        ]
    }
}
}
```
Change-Id: I4feafbbd38eaa1971abf32a37901c498e2e34b73
Signed-off-by: Aravind T. Nair <aravynd@gmail.com>

Signed-off-by: Aravind T. Nair <aravynd@gmail.com>

aravynd commented:
> please use ``` (triple backticks) to enclose test results in the PR summary.

I added triple backticks, but I dont really see a change

devenrao commented on error/sbe_hwp_failure.H:
```diff
@@ -0,0 +1,17 @@
+
+namespace errl::sbe_hwp_failure
+{
+/** @brief Create an error handle for an SBE boot failure.
```
SBE hwp failure - typo

devenrao commented on error/sbe_hwp_failure.H:
```diff
@@ -0,0 +1,17 @@
+/** @brief Create an error handle for an SBE boot failure.
+ *
+ *  Constructs and returns an error log handle specific to a Self Boot Engine
+ * (SBE) boot failure scenario on the provided hardware target.
```
fix boot to hwp

devenrao commented on error/sbe_hwp_failure.C:
```diff
@@ -0,0 +1,26 @@
+#include <libekb.H>
+#include <sbe_hwp_failure.H>
+
+#include <log.hpp>
```
remove unused

devenrao commented on libipl/p10/ipl0.C:
```diff
@@ -1382,7 +1382,7 @@ try{
         else
         {
-            //err_type = IPL_ERR_HWP;
+            ipl_error_callback(IPL_ERR_HWP);
```
not required as we are checking the retun value of ipl_run_major and not relying on callback
